tests/bls_tests.rs

```diff
@@ -0,0 +1,341 @@
+use std::fs::File;
+
+fn decode_hex_with_prefix<T: AsRef<[u8]> + Debug>(data: T) -> Vec<u8> {
+    // a small check for empty strings
+    if data.as_ref().len() >= 2 {
+        hex::decode(&data.as_ref()[2..]).expect("is well-formed hex")
+    } else {
+        hex::decode(&data.as_ref()).expect("is well-formed hex")
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct SigningInput {
+    privkey: String,
+    message: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct SigningTestIO {
+    input: SigningInput,
+    output: String,
+}
+
+impl SigningTestIO {
+    fn verify_signature(&self) -> bool {
+        let input = &self.input;
+
+        let secret_key_bytes = decode_hex_with_prefix(&input.privkey);
+        let secret_key = match SecretKey::from_bytes(&secret_key_bytes) {
+            Ok(sk) => sk,
+            Err(_) => {
+                // this is the case empty secret key case
+                if self.output == String::from("~") {
+                    return true;
+                } else {
+                    return false;
+                }
+            }
+        };
+
+        let message_bytes = decode_hex_with_prefix(&input.message);
+        let signature = secret_key.sign(&message_bytes);
+
+        let signature_bytes = decode_hex_with_prefix(&self.output);
+        let expected_signature = Signature::from_bytes(&signature_bytes).unwrap();
+
+        signature == expected_signature
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+
+struct AggregatingTestIO {
+    input: Vec<String>,
+    output: String,
+}
+
+impl AggregatingTestIO {
+    fn verify_aggregate(&self) -> bool {
+        let input_signatures: Vec<Signature> = self
+            .input
+            .iter()
+            .map(|x| Signature::from_bytes(&decode_hex_with_prefix(&x)).unwrap())
+            .collect();
+        let aggregate = match aggregate(&input_signatures) {
+            Ok(agg) => agg,
+            // handling for zero sized input and output
+            Err(_) => {
+                if self.output == String::from("~") {
+                    return true;
+                } else {
+                    return false;
+                }
+            }
+        };
+        let expected_aggregate_raw = decode_hex_with_prefix(&self.output);
+        let expected_aggregate = Signature::from_bytes(&expected_aggregate_raw).unwrap();
+        aggregate == expected_aggregate
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+
+struct AggVerifyInput {
+    pubkeys: Vec<String>,
+    messages: Vec<String>,
+    signature: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+
+struct AggVerifyTestIO {
+    input: AggVerifyInput,
+    output: bool,
+}
+
+impl AggVerifyTestIO {
+    fn aggregate_verify(&self) -> bool {
+        let pubkeys_result: Result<Vec<PublicKey>, _> = self
+            .input
+            .pubkeys
+            .iter()
+            .map(|x| PublicKey::from_bytes(&decode_hex_with_prefix(&x)))
```

Comment:
do you need this ref on `x`?
```suggestion
            .map(|x| PublicKey::from_bytes(&decode_hex_with_prefix(x)))
```